src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Refactoring/RazorComponentRenameEndpoint.cs

```diff
@@ -0,0 +1,292 @@
+
+        public async Task<WorkspaceEdit> Handle(RenameParams request, CancellationToken cancellationToken)
+        {
+            if (request is null)
+            {
+                throw new ArgumentNullException(nameof(request));
+            }
+
+            var documentSnapshot = await Task.Factory.StartNew(() =>
+            {
+                _documentResolver.TryResolveDocument(request.TextDocument.Uri.GetAbsoluteOrUNCPath(), out var documentSnapshot);
+                return documentSnapshot;
+            }, cancellationToken, TaskCreationOptions.None, _foregroundDispatcher.ForegroundScheduler).ConfigureAwait(false);
+
+            if (documentSnapshot is null)
+            {
+                return null;
+            }
+
+            var codeDocument = await documentSnapshot.GetGeneratedOutputAsync().ConfigureAwait(false);
+            if (codeDocument.IsUnsupported())
+            {
+                return null;
+            }
+
+            if (!FileKinds.IsComponent(codeDocument.GetFileKind()))
+            {
+                return null;
+            }
+
+            var originTagHelperBinding = await GetOriginTagHelperBindingAsync(documentSnapshot, codeDocument, request.Position).ConfigureAwait(false);
+            if (originTagHelperBinding is null)
+            {
+                return null;
+            }
+
+            var originTagDescriptor = originTagHelperBinding.Descriptors.First();
+            var originComponentDocumentSnapshot = await await Task.Factory.StartNew(() =>
+            {
+                return _componentSearchEngine.TryLocateComponentAsync(originTagDescriptor).ConfigureAwait(false);
+            }, cancellationToken, TaskCreationOptions.None, _foregroundDispatcher.ForegroundScheduler).ConfigureAwait(false);
+
+            if (originComponentDocumentSnapshot is null)
+            {
+                return null;
+            }
+
+            var newPath = MakeNewPath(documentSnapshot.FilePath, request.NewName);
+            if (File.Exists(newPath))
+            {
+                return null;
+            }
+
+            var documentChanges = new List<WorkspaceEditDocumentChange>();
+            AddFileRenameForComponent(documentChanges, originComponentDocumentSnapshot, newPath);
+            AddEditsForCodeDocument(documentChanges, originTagHelperBinding, request.NewName, request.TextDocument.Uri, codeDocument);
+
+            var projects = await Task.Factory.StartNew(() =>
+            {
+                return _projectSnapshotManager.Projects.ToArray();
+            }, cancellationToken, TaskCreationOptions.None, _foregroundDispatcher.ForegroundScheduler).ConfigureAwait(false);
+
+            var documentPaths = new HashSet<string>();
+            foreach (var project in projects)
+            {
+                foreach (var documentPath in project.DocumentFilePaths)
+                {
+                    if (string.Equals(documentPath, documentSnapshot.FilePath, FilePathComparison.Instance))
```

Comment:
Technically a component can recursively call itself so you'll probably still want to allow the current snapshot 😄

Reply:
This is skipping the document we're invoking the rename from which we've already added changes for (to avoid setting up the code document and everything twice). I'm gonna improve variable naming so this is more clear.

Reply:
oooo